tests/test_config.py

```diff
@@ -31,27 +37,87 @@ class AnExample1Config(ConfigBase):
 
 
+@pytest.fixture(scope="session")
+def toml_file(tmp_path_factory: pytest.TempPathFactory) -> Path:
+    file_path = (
+        tmp_path_factory.mktemp(AnExample1Config.default_foldername())
+        / AnExample1Config.default_filename()
+    )
+    with file_path.open(mode="wb") as fptr:
+        tomli_w.dump({"section1": {"field1": "f1", "field2": 22}}, fptr)
+    return file_path
+
+
+@pytest.fixture(scope="session")
+def json_file(tmp_path_factory: pytest.TempPathFactory) -> Path:
+    file_path = tmp_path_factory.mktemp(
+        AnExample1Config.default_foldername()
+    ) / AnExample1Config.default_filename().replace("toml", "json")
+    with file_path.open(mode="w") as fptr:
+        json.dump({"section1": {"field1": "f2", "field2": 33}}, fptr)
+    return file_path
+
+
+@pytest.fixture(scope="session")
+def ini_file(tmp_path_factory: pytest.TempPathFactory) -> Path:
+    file_path = tmp_path_factory.mktemp(
+        AnExample1Config.default_foldername()
+    ) / AnExample1Config.default_filename().replace("toml", "ini")
+    with file_path.open(mode="w") as fptr:
+        fptr.write("inifile")
+    return file_path
+
+
 def test_version() -> None:
-    assert __version__ == "0.1.0"
+    assert __version__ == "0.2.0"
 
 
-def test_defaults(
-    monkeypatch: pytest.MonkeyPatch, capfd: pytest.CaptureFixture[str]
-) -> None:
-    the_path = AnExample1Config.default_config_filepath()
-    if the_path:
+def test_paths(toml_file: Path) -> None:
+    # default_filepath:
+    if the_path := AnExample1Config.default_filepath():
         assert the_path.parts[-1] == "config.toml"
         assert the_path.parts[-2] == ".an_example1"
-
+    else:
+        assert False
+
+    # filepath:
+    # if not set, then equal to default_filepath
+    assert AnExample1Config.filepath() == the_path
+    # check set_filepath with a Path
+    AnExample1Config.set_filepath(toml_file)
+    assert AnExample1Config.filepath() == toml_file
+    # check set_filepath with a str
+    AnExample1Config.set_filepath(".")
+    assert AnExample1Config.filepath() == Path.cwd()
+
+    # reset to default:
+    AnExample1Config.set_filepath("")
+    assert AnExample1Config.filepath() == the_path
+
+    # raising of FileNotFoundError:
     with pytest.raises(FileNotFoundError):
         _ = AnExample1Config.get().section1.field1
 
-    def mock_default_config_filepath() -> Path | None:
-        return None
+    # raising in case of invalid path:
+    with pytest.raises(ValueError):
+        AnExample1Config.set_filepath('fi:\0\\l*e/p"a?t>h|.t<xt')
 
-    monkeypatch.setattr(
-        AnExample1Config, "default_config_filepath", mock_default_config_filepath
-    )
+
+def test_get_defaults(
+    monkeypatch: pytest.MonkeyPatch, capfd: pytest.CaptureFixture[str]
+) -> None:
+    if sys.version_info >= (3, 10):
+
+        def mock_default_filepath() -> Path | None:
+            return None
+
+    else:
+
+        def mock_default_filepath() -> Union[Path, None]:
```

Comment:
As mentioning we might just want to export these types in a typevar that are used more often in a separate file. That avoids a lot of the code duplication hopefully.

Reply:
Done